In [21]:
import pandas as pd
df=pd.read_csv("Absenteeism_preprocessed.csv")

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 15 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Group_1                    700 non-null    int64  
 1   Group_2                    700 non-null    int64  
 2   Group_3                    700 non-null    int64  
 3   Group_4                    700 non-null    int64  
 4   month                      700 non-null    int64  
 5   day of the week            700 non-null    int64  
 6   Transportation Expense     700 non-null    int64  
 7   Distance to Work           700 non-null    int64  
 8   Age                        700 non-null    int64  
 9   Daily Work Load Average    700 non-null    float64
 10  Body Mass Index            700 non-null    int64  
 11  Education                  700 non-null    int64  
 12  Children                   700 non-null    int64  
 13  Pets                       700 non-null    int64  

# Create logistic regression for analysis

# Analysis of column from user side

1. reason columns may be imortant i.e Group 1,2,3,4
2. daily work load may cause skiping the work day
3. having childern or pet sick may cause absenteeism

In [23]:
df.head()

Group_1  Group_2  Group_3  Group_4  month  day of the week  \
0        0        0        0        1      7                1   
1        0        0        0        0      7                1   
2        0        0        0        1      7                2   
3        1        0        0        0      7                3   
4        0        0        0        1      7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          0         2     1                          4  
1               31          0         1     0                          0  
2               31          0         0     0                          2  
3               24          0         2     0                          4  
4               30          0         2     1                          2

everything below the mdeian considered as normal above the median can be excessive data or outlier in other word

In [24]:
# lets checkout for absenteeism hours and its median value
df['Absenteeism Time in Hours'].median()

3.0

### conclusion being absent 3 hrs is normal but more than that affects the work

In [25]:
# lets convert the data 3 = 0 and more then 3 as 1 in absenteeism hours
import numpy as np
target = np.where(df['Absenteeism Time in Hours']>3,1,0)

In [26]:
target

array([1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,

In [27]:
# more precise way to convet the data to 0 and 1  ot minimise the mistakes
target = np.where(df['Absenteeism Time in Hours']>df['Absenteeism Time in Hours'].median(),1,0)

In [28]:
target

array([1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,

In [29]:
# add the new col to data as target or excessive absenteeism
df['excessive ansenteeism']=target

In [31]:
df_target = df.drop(['Absenteeism Time in Hours'],axis=1)

In [32]:
df_target.columns.values

array(['Group_1', 'Group_2', 'Group_3', 'Group_4', 'month',
       'day of the week', 'Transportation Expense', 'Distance to Work',
       'Age', 'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets', 'excessive ansenteeism'], dtype=object)

In [33]:
# check the data is balanced or not for target column (sum of all the 1/ total num of rows)

target.sum()/target.shape[0]

0.45571428571428574

# we can say 45% of data for 1 and remaining 55% data is of 0=our data is quiet balanced and we can proceed with the same

In [34]:
# now selact your x and y
unscaled_X = df_target.iloc[:,:-1]  # unscaled_X = df_target.iloc[:,0:14] or [:,:14]

# standardize the data

In [36]:
from sklearn.preprocessing import StandardScaler
scale = StandardScaler()
scaled_x = scale.fit(unscaled_X)
scaled_x = scale.transform(unscaled_X)

In [45]:
scaled_x.shape
y= df_target['excessive ansenteeism']

In [46]:
from sklearn.model_selection import train_test_split

In [51]:
X_train,X_test,y_train,y_test=train_test_split(scaled_x,y,random_state=20,test_size=0.2)

In [49]:
print(X_train.shape,y_train.shape)

(560, 14) (560,)


In [50]:
print(X_test.shape,y_test.shape)

(140, 14) (140,)


In [53]:
# model importing and accuracy chack
from sklearn.linear_model import LogisticRegression
reg = LogisticRegression()
reg.fit(X_train,y_train)
reg.score(X_train,y_train)

0.7839285714285714

In [54]:
# coefficicients
reg.coef_

array([[ 2.07601767,  0.33504757,  1.56162303,  1.32927434,  0.18793677,
        -0.07062253,  0.70639316, -0.03986811, -0.20089491, -0.00456366,
         0.31933564, -0.135508  ,  0.38172443, -0.3332426 ]])

In [55]:
unscaled_X.columns.values

array(['Group_1', 'Group_2', 'Group_3', 'Group_4', 'month',
       'day of the week', 'Transportation Expense', 'Distance to Work',
       'Age', 'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets'], dtype=object)

In [60]:
feature_names = unscaled_X.columns.values
# create the coeff of feature and there values in dataframe
summary = pd.DataFrame(columns=['Feature_names'],data=feature_names)
summary['coefficients']= np.transpose(reg.coef_)
summary

Feature_names  coefficients
0                   Group_1      2.076018
1                   Group_2      0.335048
2                   Group_3      1.561623
3                   Group_4      1.329274
4                     month      0.187937
5           day of the week     -0.070623
6    Transportation Expense      0.706393
7          Distance to Work     -0.039868
8                       Age     -0.200895
9   Daily Work Load Average     -0.004564
10          Body Mass Index      0.319336
11                Education     -0.135508
12                 Children      0.381724
13                     Pets     -0.333243

In [61]:
# intercept is missing in summary table
reg.intercept_

array([-0.22206736])

In [62]:
# intercept is one value we can add it in table abt zoroth position and coeff from 1st inexx values

In [63]:
summary.index= summary.index+1

In [64]:
summary

Feature_names  coefficients
1                   Group_1      2.076018
2                   Group_2      0.335048
3                   Group_3      1.561623
4                   Group_4      1.329274
5                     month      0.187937
6           day of the week     -0.070623
7    Transportation Expense      0.706393
8          Distance to Work     -0.039868
9                       Age     -0.200895
10  Daily Work Load Average     -0.004564
11          Body Mass Index      0.319336
12                Education     -0.135508
13                 Children      0.381724
14                     Pets     -0.333243

In [65]:
summary.loc[0]=['intercept',reg.intercept_[0]]

In [66]:
summary

Feature_names  coefficients
1                   Group_1      2.076018
2                   Group_2      0.335048
3                   Group_3      1.561623
4                   Group_4      1.329274
5                     month      0.187937
6           day of the week     -0.070623
7    Transportation Expense      0.706393
8          Distance to Work     -0.039868
9                       Age     -0.200895
10  Daily Work Load Average     -0.004564
11          Body Mass Index      0.319336
12                Education     -0.135508
13                 Children      0.381724
14                     Pets     -0.333243
0                 intercept     -0.222067

In [69]:
# now adding intercept as 0 index we need to sort it 
summary = summary.sort_index()

In [70]:
summary

Feature_names  coefficients
0                 intercept     -0.222067
1                   Group_1      2.076018
2                   Group_2      0.335048
3                   Group_3      1.561623
4                   Group_4      1.329274
5                     month      0.187937
6           day of the week     -0.070623
7    Transportation Expense      0.706393
8          Distance to Work     -0.039868
9                       Age     -0.200895
10  Daily Work Load Average     -0.004564
11          Body Mass Index      0.319336
12                Education     -0.135508
13                 Children      0.381724
14                     Pets     -0.333243

In [71]:
# coefficients = weights(log(odds)) and intercept=bias
# we have standardized coefficients as we have standardized our data so the varience will nbe one
# more the weight/coefficient the corresponding feature is more important does not matter positive or negative

# find the odds ration i.e we need to take exponential of coefficients

In [75]:
summary['odds_ratio']=np.exp(summary['coefficients'])
summary.sort_values('odds_ratio',ascending=False)

Feature_names  coefficients  odds_ratio
1                   Group_1      2.076018    7.972656
3                   Group_3      1.561623    4.766551
4                   Group_4      1.329274    3.778301
7    Transportation Expense      0.706393    2.026668
13                 Children      0.381724    1.464808
2                   Group_2      0.335048    1.398007
11          Body Mass Index      0.319336    1.376213
5                     month      0.187937    1.206757
10  Daily Work Load Average     -0.004564    0.995447
8          Distance to Work     -0.039868    0.960916
6           day of the week     -0.070623    0.931814
12                Education     -0.135508    0.873272
9                       Age     -0.200895    0.817998
0                 intercept     -0.222067    0.800861
14                     Pets     -0.333243    0.716596

# reading of coeff and odds ratio
1. if coeff value near to 0 or exact 0 and odds ratio is 1 the corresponding feature is not important
example = Daily Work Load Average coeff = 0.004564 and odds ratio = 0.995447 
the above feature is not important

similarly : Distance to Work, day of the week are also not the important features

### very important check point

whenever we scaled our data we scalled our dummy variables to which is not the correct option

while scalling the data we should skip or drop the dummies while scalling but we need to consider them as your features while fiting the algorithm and train test data

In [78]:
unscaled_X.columns.values

array(['Group_1', 'Group_2', 'Group_3', 'Group_4', 'month',
       'day of the week', 'Transportation Expense', 'Distance to Work',
       'Age', 'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets'], dtype=object)

In [79]:
# select those cli that we need to scale i.e we need to drop dummies 
col_scale = ['month','day of the week', 'Transportation Expense', 'Distance to Work',
       'Age', 'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets']

In [ ]:
scale = StandardScaler()
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.preprocessing import StandardScaler
# code to drop dummies from standardization process
class CustomScaler(BaseEstimator,TransformerMixin): 
    
    def __init__(self,columns,copy=True,with_mean=True,with_std=True):
        self.scaler = StandardScaler(copy,with_mean,with_std)
        self.columns = columns
        self.mean_ = None
        self.var_ = None

    def fit(self, X, y=None):
        self.scaler.fit(X[self.columns], y)
        self.mean_ = np.mean(X[self.columns])
        self.var_ = np.var(X[self.columns])
        return self

    def transform(self, X, y=None, copy=None):
        init_col_order = X.columns
        X_scaled = pd.DataFrame(self.scaler.transform(X[self.columns]), columns=self.columns)
        X_not_scaled = X.loc[:,~X.columns.isin(self.columns)]
        return pd.concat([X_not_scaled, X_scaled], axis=1)[init_col_order]
